File: python/src/validators.py
"""Validation des entrees utilisateur."""

from datetime import date
from .exceptions import InvalidInputError

VALID_PRIORITIES = {"low", "medium", "high"}
VALID_STATUSES = {"todo", "doing", "done"}

TITLE_MIN = 1
TITLE_MAX = 100
DESCRIPTION_MAX = 500


def validate_title(value):
    if not isinstance(value, str):
        raise InvalidInputError("Le titre doit etre une chaine de caracteres.")
    cleaned = value.strip()
    if len(cleaned) < TITLE_MIN:
        raise InvalidInputError("Le titre ne peut pas etre vide.")
    if len(cleaned) > TITLE_MAX:
        raise InvalidInputError(f"Le titre doit faire au plus {TITLE_MAX} caracteres.")
    return cleaned


def validate_description(value):
    if value is None:
        return ""
    if not isinstance(value, str):
        raise InvalidInputError("La description doit etre une chaine de caracteres.")
    cleaned = value.strip()
    if len(cleaned) > DESCRIPTION_MAX:
        raise InvalidInputError(f"La description doit faire au plus {DESCRIPTION_MAX} caracteres.")
    return cleaned


def validate_priority(value):
    if value not in VALID_PRIORITIES:
        raise InvalidInputError(
            f"Priorite invalide '{value}'. Valeurs attendues : {sorted(VALID_PRIORITIES)}."
        )
    return value


def validate_status(value):
    if value not in VALID_STATUSES:
        raise InvalidInputError(
            f"Statut invalide '{value}'. Valeurs attendues : {sorted(VALID_STATUSES)}."
        )
    return value


def validate_due_date(value):
    if value is None or value == "":
        return None
    if not isinstance(value, str):
        raise InvalidInputError("La date doit etre une chaine au format AAAA-MM-JJ.")
    try:
        return date.fromisoformat(value).isoformat()
    except ValueError:
        raise InvalidInputError(
            f"Date invalide '{value}'. Format attendu : AAAA-MM-JJ (ex : 2026-12-31)."
        )

File: typescript/src/validators.ts
import { InvalidInputError } from "./errors.js";

export type Priority = "low" | "medium" | "high";
export type Status = "todo" | "doing" | "done";

export const VALID_PRIORITIES: Priority[] = ["low", "medium", "high"];
export const VALID_STATUSES: Status[] = ["todo", "doing", "done"];
export const TITLE_MIN = 1;
export const TITLE_MAX = 100;
export const DESCRIPTION_MAX = 500;

export function validateTitle(value: unknown): string {
  if (typeof value !== "string") {
    throw new InvalidInputError("Le titre doit etre une chaine de caracteres.");
  }
  const cleaned = value.trim();
  if (cleaned.length < TITLE_MIN) {
    throw new InvalidInputError("Le titre ne peut pas etre vide.");
  }
  if (cleaned.length > TITLE_MAX) {
    throw new InvalidInputError(`Le titre doit faire au plus ${TITLE_MAX} caracteres.`);
  }
  return cleaned;
}

export function validateDescription(value: unknown): string {
  if (value === undefined || value === null) return "";
  if (typeof value !== "string") {
    throw new InvalidInputError("La description doit etre une chaine de caracteres.");
  }
  const cleaned = value.trim();
  if (cleaned.length > DESCRIPTION_MAX) {
    throw new InvalidInputError(`La description doit faire au plus ${DESCRIPTION_MAX} caracteres.`);
  }
  return cleaned;
}

export function validatePriority(value: unknown): Priority {
  if (!VALID_PRIORITIES.includes(value as Priority)) {
    throw new InvalidInputError(
      `Priorite invalide '${value}'. Valeurs attendues : ${VALID_PRIORITIES.join(", ")}.`
    );
  }
  return value as Priority;
}

export function validateStatus(value: unknown): Status {
  if (!VALID_STATUSES.includes(value as Status)) {
    throw new InvalidInputError(
      `Statut invalide '${value}'. Valeurs attendues : ${VALID_STATUSES.join(", ")}.`
    );
  }
  return value as Status;
}

export function validateDueDate(value: unknown): string | null {
  if (value === undefined || value === null || value === "") return null;
  if (typeof value !== "string") {
    throw new InvalidInputError("La date doit etre une chaine au format AAAA-MM-JJ.");
  }
  const isoRegex = /^\d{4}-\d{2}-\d{2}$/;
  if (!isoRegex.test(value)) {
    throw new InvalidInputError(
      `Date invalide '${value}'. Format attendu : AAAA-MM-JJ (ex : 2026-12-31).`
    );
  }
  const d = new Date(value + "T00:00:00Z");
  if (Number.isNaN(d.getTime()) || d.toISOString().slice(0, 10) !== value) {
    throw new InvalidInputError(`Date invalide '${value}'.`);
  }
  return value;
}
